devices/check.py

```diff
@@ -0,0 +1,205 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import hashlib
+import json
+import os
+import sys
+import subprocess
+import tempfile
+
+class CheckFirmware:
+    def __init__(self, fwanalyzer="fwanalyzer"):
+        self._tmpdir = tempfile.mktemp()
+        self._unpackdir = os.path.join(self._tmpdir, "unpacked")
+        self._fwanalyzer = fwanalyzer
+        self._unpacked = False
+
+    def get_tmp_dir(self):
+        return self._tmpdir
+
+    def run_fwanalyzer_fs(self, img, cfg, cfginc, out, options=""):
+        cfginclude = ""
+        if cfginc:
+            cfginclude = " -cfgpath {0}".format(cfginc)
+        cmd = "{0} -in {1} {2} -cfg {3} -out {4} {5}".format(self._fwanalyzer, img, cfginclude, cfg, out, options)
+        return subprocess.check_call(cmd, shell=True)
+
+    def unpack(self, fwfile, unpacker, cfgpath):
+        try:
+            if os.path.exists(os.path.join(fwfile, "unpacked")) and os.path.exists(os.path.join(fwfile, "targets.json")):
+                self._tmpdir = fwfile
+                self._unpackdir = os.path.join(self._tmpdir, "unpacked")
+                print("{0}: is a directory containing an 'unpacked' path, skipping".format(fwfile))
+                cmd = "cat {0}".format(os.path.join(fwfile, "targets.json"))
+                self._unpacked = True
+            else:
+                os.mkdir(self._tmpdir)
+                os.mkdir(self._unpackdir)
+                cmd = "{0} {1} {2}".format(unpacker, fwfile, cfgpath)
+            res = subprocess.check_output(cmd, shell=True, cwd=self._unpackdir)
+            targets = json.loads(res.decode('utf-8'))
+            with open(os.path.join(self._tmpdir, "targets.json"), "w+") as fp:
+                fp.write(res.decode('utf-8'))
+            return targets
+        except Exception as e:
+            print("Exception: {0}".format(e))
+            print("can't load targets from output of '{0}' check your script".format(unpacker))
+            return None
+
+    def del_tmp_dir(self):
+        if not self._unpacked:
+            cmd = "rm -rf {0}".format(self._tmpdir)
+            return subprocess.check_call(cmd, shell=True)
+
+    def files_by_ext_stat(self, data):
+        allext = {}
+        for i in data["files"]:
+            fn, ext = os.path.splitext(i["name"])
+            if ext in allext:
+                count, ext = allext[ext]
+                allext[ext] = count + 1, ext
+            else:
+                allext[ext] = (1, ext)
+        return (len(data["files"]), allext)
+
+    def analyze_filetree(self, filetreefile):
+        with open(filetreefile) as fp:
+            data = json.load(fp)
+        num_files, stats = self.files_by_ext_stat(data)
+        out = {}
+        percent = num_files / 100
```

Comment:
does this work? if num_files is an int then percent will end up being 0 if num_files is < 100

Reply:
yes and then we get stats for every file extension :)